---
title: "GEViTRec"
author: "Anamaria Crisan"
date: '2019-02-01'
output: html_document
---


## Loading data
```{r}
devtools::load_all() #temporary once things are done

#Table data
tab_dat<-input_data(file = system.file("./inst/extdata/", "ebov_metadata.csv", package = "epivis"),dataType = "table")

#Tree data
tree_dat<-input_data(file = system.file("./inst/extdata/", "ebov_tree.nwk", package = "epivis"),dataType = "tree")

#Genomic data
genomic_dat<-input_data(file = system.file("./inst/extdata/", "ebov_GIN_genomic.fasta", package = "epivis"),dataType = "dna")

#Shape files
#Shape files require that .shp,.shx,and .prj files at a minimun to be in the same directory
#to add metadata to the shape file, you can also add .dbf files
gin_shape_dat<-input_data(file = system.file("./inst/extdata/", "gin_admbnda_adm1_ocha_itos.shp", package = "epivis"),dataType = "spatial")
lbr_shape_dat<-input_data(file = system.file("./inst/extdata/", "lbr_admbnda_adm1_ocha.shp", package = "epivis"),dataType = "spatial")
sle_shape_dat<-input_data(file = system.file("./inst/extdata/", "sle_admbnda_adm1_1m_gov_ocha_20161017.shp",package = "epivis"),dataType = "spatial")
```

## Data Harmonization
```{r}
devtools::load_all()
harmon_obj<-data_harmonization(tab_dat,tree_dat,genomic_dat,gin_shape_dat,lbr_shape_dat,sle_shape_dat)
```

### Show how data objects are related to each other
```{r,fig.height=7,units="in"}
view_entity_graph(harmon_obj[["entityGraph"]])

view_entity_graph(subset_graph(harmon_obj[["entityGraph"]],1))
```

## Chart Recommendations

### Temporary - sorting out few functions
```{r}
objMeta<-harmon_obj[["dataMeta"]]
entity_graph<-subset_graph(harmon_obj[["entityGraph"]],1)

entity_graph<-harmon_obj[["entityGraph"]]

entity_graph_table<-as_tibble(entity_graph)

usrChoices<-c() #the user can state which variables should be in the visualization

#number of data types per component
component_info<- entity_graph_table %>%
  dplyr::filter(dataEntity == "dataType") %>%
  dplyr::group_by(component) %>%
  dplyr::count() %>%
  dplyr::arrange(desc(n))

for(comp in component_info$component){
  comp_info<-dplyr::filter(component_info,component == comp)
  
  #for each component, try to produce at most two
  #visualizations
  max_vis<-2
  if(comp_info$n>1){
    #the hard coded numbers here are abitrary
    max_vis<-ifelse(comp_info$n*2<10,10,comp_info$n*2)
  }
  
  #get the relevant data for each component
  comp_var<-dplyr::filter(entity_graph_table,dataEntity=="feild") %>%
    dplyr::filter(component == comp)
  
  comp_data<-dplyr::filter(entity_graph_table,dataEntity=="dataType") %>%
    dplyr::filter(component == comp)
  
  #Find paths between variables as these are essential
  #seeds for the specifications
  dats<-as.numeric(V(entity_graph)[as.character(comp_data$name)])
  dat_paths<-igraph::all_shortest_paths(entity_graph,from=dats,to=dats)$res
  
  for(idx in 1:length(dat_paths)){
    conn_nodes<-as_ids(dat_paths[[1]])
    seed_specs<-dplyr::filter(entity_graph_table,name %in% conn_nodes)
    gen_specs(seed_specs)
  }
}

```

Using the example below, there are a total of 300 possible combinations of specifications, but when we apply the constraints on the graph for nodes with high degrees that connect information, there are now suddenly only 9 specifications that we care about. MUCH MUCH more mangeable.

Can futher narrow it down if the user has provided variables they care about, then you only want those kinds of combinations - but, actually would be good to show both and just rank the user required ones higher

Also - write out the math for this to prove this assertion because it will help in the paper

```{r}
datFeilds<-filter(entity_graph_table,dataSource == "table_VTFEH2385H") %>%
  filter(dataEntity == "feild")

require_var="longitude"

filter_combo<-function(combos,require_var=NULL){
  if(is.null(require_var))
    return(combos)
  
    req_present<-apply(combos,2,function(x,req_var){
       sum(req_var %in% x)
    },req_var = require_var)
    
    n_req<-nrow(combos) #default for combn output
   
     if(length(require_var)<n_req){ n_req <-length(require_var)}
    
    idx_keep<-req_present>=n_req
    return(combos[,idx_keep])
}

assign_vars<-function(data = NULL,datFeilds = NULL,require_var = NULL,n_quant=3,n_qual=2){
  #high_degree vars have priority, because they can
  #faciliate combinations with other data
  connect_node<-dplyr::filter(datFeilds, degree>1) %>%
    dplyr::arrange(desc(degree)) %>%
    head(5)  
    
  require_var<-c(require_var,connect_node$name)
    
  quant<-dplyr::filter(datFeilds,dataType %in% c("double"," integer"))
  qual<-dplyr::filter(datFeilds,dataType %in% c("character"))
  
  quant_require<- if (!is.null(require_var)) intersect(quant$name,require_var) else NULL
  qual_require<- if (!is.null(require_var)) intersect(qual$name,require_var) else NULL
    
  #a bit exaggerated because it doesn't take require variales into account
  n_combos<- choose(length(qual),n_qual) * choose(length(qual),n_quant)
  
  #and goo..
  if(n_combos>1000){
    stop("Haven't implemented this yet")
  }
  
  combo_quant<-filter_combo(combn(quant$name,m=n_quant),require_var = quant_require)
  combo_qual<-filter_combo(combn(qual$name,m=n_qual),require_var = qual_require)

  #generate specifications
  spec_list<-apply(combo_quant,2,function(x,combo_qual,n_quant,n_qual){
    spec<-apply(combo_qual,2,function(y,x,n_quant,n_qual){
      quant_vec<-rep(NA,n_quant)
      qual_vec<-rep(NA,n_qual)
        
      quant_vec[1:length(x)]<-x
      qual_vec[1:length(y)]<-y
        
      spec<-c(quant_vec,qual_vec)
      list(spec)
    },x = x,n_quant = n_quant,n_qual = n_qual)
    spec<-unlist(spec,recursive = FALSE)
  },combo_qual=combo_qual) %>%
      lapply(.,function(x){
        do.call(rbind,x)
      }) %>% do.call(rbind,.)
  
  return(spec_list)
}

```

For most chart types, positional variables will already be take up (like a phylogenetic tree), and so the quantiative positional variables don't matter, but non-positional channels can be added via other variables if the chart type has its own metadata, or is linked to another variable that does with a direct link.

### Temporary - loading and preping gevit data
```{r}
gevit_articles<-readxl::read_xlsx(path="../../../gevit_gallery_v2/data/MasterDocumentList.xlsx")
gevit_taxonomy<-readxl::read_xlsx(path="../../../gevit_gallery_v2/data/figure_classification_final.xlsx")
```